sparkly/session.py

```diff
@@ -91,21 +90,24 @@ def __init__(self, additional_options=None):
             self._setup_options(additional_options),
             'pyspark-shell',
         ]
-        os.environ['PYSPARK_SUBMIT_ARGS'] = ' '.join(filter(None, submit_args))
-
-        # If we are in instant testing mode
-        if InstantTesting.is_activated():
-            spark_context = InstantTesting.get_context()
 
-            # It's the first run, so we have to create context and demonise the process.
-            if spark_context is None:
+        with temporary_dict_copy(os.environ):
```

Comment:
Because `SparlySession` will modify env `PYSPARK_SUBMIT_ARGS`. We have to set it back to original state when we initialize a new session.

Reply:
Yep, in order to truly support multiple sparkly sessions with say different package requirements during testing, each session has to have access to the same exact environment that the previous one had when it gets initialized, since each session honors the previously set `PYSPARK_SUBMIT_ARGS`, and just appends to it. Otherwise, the second session we try to initialize will try to honor the modified version of `PYSPARK_SUBMIT_ARGS` from the first session, which will include options like `--packages` etc. that were set just for the needs of the first session, and there will be argument conflicts at best, or errors at worst.

So as a generic solution, we make sure that `SparklySession` (and the initialization of `SparkContext`, which is what controls the gateway / jars / etc) leaves no trace on the environment. The env is clean and ready for the next session to use.

Reply:
I am a bit concerned with this solution cause we don't know all the places where Spark relies on the env variables set during `SparkContext` initialisation. E.g. we initialise `HiveContext` outside the `with temporary_dict_copy(os.environ):` block. What do you think guys?

Reply:
@drudim Even though the tests pass, we don't really cover everything that spark can do, so that's a fair concern. How about we copy the state of `os.environ` on `__init__` and then restore it on `stop`? This way we wouldn't mess with any environment variables that spark might need while it's running, but still cleanup appropriately when we're finished with one session.